File: 135-candy/135-candy.py
class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        n = len(ratings)
        l = [1 for i in range(n)]
        r = [1 for j in range(n)]
        candies = 0
        for k in range(1, n, 1):
            if(ratings[k] > ratings[k-1]):
                l[k] = l[k-1] + 1
        
        for m in range(n-2, -1, -1):
            if(ratings[m] > ratings[m+1]):
                r[m] = r[m+1] + 1
                
        for x in range(n):
            candies = candies + max(l[x], r[x])
            
        return candies

File: 135-candy/135-candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int candies = 0;
        int up = 0, down = 0;
        int prevslope = 0;
        for(int i=1; i<n; i++){
            int currslope = ratings[i] > ratings[i-1] ? 1
                : (ratings[i] < ratings[i-1] ? -1:0);
            
            if((prevslope<0 && currslope>=0)||(prevslope>0 && currslope==0)){
                candies = candies + sum(up) + sum(down) + Math.max(up, down);
                up = 0;
                down = 0;
            }
            
            if(currslope > 0) up++;
            else if(currslope < 0) down++;
            else candies++;
            
            prevslope = currslope;
        }
        candies = candies + sum(up) + sum(down) + Math.max(up, down) + 1;
        return candies;
    }
    private int sum(int n){
        return (n*(n+1))/2;
    }
    
}